util/tracecpu/profile.go

```diff
@@ -0,0 +1,437 @@
+	var task *profileTask
+	select {
+	case task = <-sp.cacheBufCh:
+		task.buf.Reset()
+	default:
+		task = &profileTask{
+			buf: bytes.NewBuffer(make([]byte, 0, 100*1024)),
+		}
+	}
+	return task
+}
+
+func (sp *sqlStatsProfiler) putTaskToBuffer(task *profileTask) {
+	select {
+	case sp.cacheBufCh <- task:
+	default:
+	}
+}
+
+// parseCPUProfileBySQLLabels uses to aggregate the cpu-profile sample data by sql_digest and plan_digest labels,
+// output the SQLStats slice. Want to know more information about profile labels, see https://rakyll.org/profiler-labels/
+// The sql_digest label is been set by `SetGoroutineLabelsWithSQL` function after parse the SQL.
+// The plan_digest label is been set by `SetGoroutineLabelsWithSQLAndPlan` function after build the SQL plan.
+// Since `sqlStatsProfiler` only care about the cpu time that consume by (sql_digest,plan_digest), the other sample data
+// without those label will be ignore.
+func (sp *sqlStatsProfiler) parseCPUProfileBySQLLabels(p *profile.Profile) []SQLStats {
+	sqlMap := make(map[string]*sqlStats)
+	idx := len(p.SampleType) - 1
+	for _, s := range p.Sample {
+		digests, ok := s.Label[labelSQLDigest]
+		if !ok || len(digests) == 0 {
+			continue
+		}
+		for _, digest := range digests {
+			stmt, ok := sqlMap[digest]
+			if !ok {
+				stmt = &sqlStats{
+					plans: make(map[string]int64),
+					total: 0,
+				}
+				sqlMap[digest] = stmt
+			}
+			stmt.total += s.Value[idx]
+
+			plans := s.Label[labelPlanDigest]
+			for _, plan := range plans {
+				stmt.plans[plan] += s.Value[idx]
+			}
+		}
+	}
+	return sp.createSQLStats(sqlMap)
+}
+
+func (sp *sqlStatsProfiler) createSQLStats(sqlMap map[string]*sqlStats) []SQLStats {
+	stats := make([]SQLStats, 0, len(sqlMap))
+	for sqlDigest, stmt := range sqlMap {
+		stmt.tune()
+		for planDigest, val := range stmt.plans {
+			stats = append(stats, SQLStats{
+				SQLDigest:  sqlDigest,
+				PlanDigest: planDigest,
+				CPUTimeMs:  uint32(time.Duration(val).Milliseconds()),
+			})
+		}
+	}
+	return stats
+}
+
+type sqlStats struct {
+	plans map[string]int64
+	total int64
+}
+
+// tune use to adjust sql stats. Consider following situation:
+// The `sqlStats` maybe:
+//     plans: {
+//         "table_scan": 200ms, // The cpu time of the sql that plan with `table_scan` is 200ms.
+//         "index_scan": 300ms, // The cpu time of the sql that plan with `table_scan` is 300ms.
+//       },
+//     total:      600ms,       // The total cpu time of the sql is 600ms.
+// total_time - table_scan_time - index_scan_time = 100ms, and this 100ms means those sample data only contain the
+// sql_digest label, doesn't contain the plan_digest label. This is cause by the `pprof profile` is base on sample.
+// After this tune function, the `sqlStats` become to:
+//     plans: {
+//         "table_scan": 240ms, // 200 + (200/(200+300))*100
+//         "index_scan": 360ms, // 300 + (300/(200+300))*100
+//       },
+//     total:      600ms,
+func (s *sqlStats) tune() {
```

Comment:
It can only fix 1 second incorrection, instead of totally avoid stats without a plan, right? (as long as planing / compiling stage takes more than 1s)

Reply:
Yes, If there is no plan label, then TopSQL out put only know the SQL cost some CPU time, but doesn't know the plan.